File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081336Z_claude_claude-opus-5_go-stdlib/audit.go
package main

import "net/http"

// Audit and export views over stored campaign state.
//
// Both endpoints are read-only aggregations: they count what is already in the
// campaign rather than recording anything, so they add no fields to the store
// and never call flush(). Counts come straight from the in-memory slices, whose
// insertion order and contents are the same after a restart, which is what
// makes the answers deterministic.

// exportSchemaVersion identifies the shape of the export document. It is the
// export's own contract version and is deliberately independent of the SQLite
// schemaVersion in store.go: the persisted layout can change without changing
// what an export looks like to a client.
const exportSchemaVersion = 1

type campaignAuditResponse struct {
	CampaignID string `json:"campaign_id"`
	Events     int    `json:"events"`
	Quests     int    `json:"quests"`
	NPCs       int    `json:"npcs"`
	Sessions   int    `json:"sessions"`
}

type campaignExportResponse struct {
	CampaignID     string `json:"campaign_id"`
	Name           string `json:"name"`
	Characters     int    `json:"characters"`
	Quests         int    `json:"quests"`
	NPCs           int    `json:"npcs"`
	InventoryItems int    `json:"inventory_items"`
	Sessions       int    `json:"sessions"`
	SchemaVersion  int    `json:"schema_version"`
}

// ---------- GET /v1/campaigns/{id}/audit ----------

func handleCampaignAudit(w http.ResponseWriter, r *http.Request) {
	campaigns.mu.Lock()
	defer campaigns.mu.Unlock()
	c, ok := campaigns.campaigns[r.PathValue("id")]
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	writeJSON(w, http.StatusOK, campaignAuditResponse{
		CampaignID: c.ID,
		Events:     len(c.Events),
		Quests:     len(c.Quests),
		NPCs:       len(c.NPCs),
		Sessions:   len(c.Sessions),
	})
}

// ---------- GET /v1/campaigns/{id}/export ----------

func handleCampaignExport(w http.ResponseWriter, r *http.Request) {
	campaigns.mu.Lock()
	defer campaigns.mu.Unlock()
	c, ok := campaigns.campaigns[r.PathValue("id")]
	if !ok {
		writeError(w, http.StatusNotFound, "campaign not found")
		return
	}
	writeJSON(w, http.StatusOK, campaignExportResponse{
		CampaignID:     c.ID,
		Name:           c.Name,
		Characters:     len(c.Characters),
		Quests:         len(c.Quests),
		NPCs:           len(c.NPCs),
		InventoryItems: len(c.Inventory),
		Sessions:       len(c.Sessions),
		SchemaVersion:  exportSchemaVersion,
	})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T061539Z_claude_claude-sonnet-5_rust-stdlib/src/audit.rs
//! Deterministic audit log and export summaries for a campaign. Both
//! endpoints aggregate counts already tracked by [`crate::campaigns`],
//! [`crate::quests`], [`crate::npcs`], [`crate::sessions`], and
//! [`crate::inventory`] rather than introducing new storage.

use std::net::TcpStream;

use crate::campaigns;
use crate::http::respond;
use crate::inventory;
use crate::json::escape_json_string;
use crate::npcs;
use crate::quests;
use crate::sessions;

const SCHEMA_VERSION: i64 = 1;

pub(crate) fn handle_audit(stream: &mut TcpStream, campaign_id: &str) -> std::io::Result<()> {
    let (_, _, events) = match campaigns::campaign_summary(campaign_id) {
        Some(summary) => summary,
        None => return respond(stream, 404, r#"{"error":"campaign not found"}"#),
    };

    let quests_count = quests::quest_count(campaign_id);
    let npcs_count = npcs::npc_count(campaign_id);
    let sessions_count = sessions::session_count(campaign_id);

    let out = format!(
        r#"{{"campaign_id":"{}","events":{},"quests":{},"npcs":{},"sessions":{}}}"#,
        escape_json_string(campaign_id),
        events,
        quests_count,
        npcs_count,
        sessions_count
    );

    respond(stream, 200, &out)
}

pub(crate) fn handle_export(stream: &mut TcpStream, campaign_id: &str) -> std::io::Result<()> {
    let (name, characters, _) = match campaigns::campaign_summary(campaign_id) {
        Some(summary) => summary,
        None => return respond(stream, 404, r#"{"error":"campaign not found"}"#),
    };

    let quests_count = quests::quest_count(campaign_id);
    let npcs_count = npcs::npc_count(campaign_id);
    let inventory_items = inventory::inventory_item_count(campaign_id);
    let sessions_count = sessions::session_count(campaign_id);

    let out = format!(
        r#"{{"campaign_id":"{}","name":"{}","characters":{},"quests":{},"npcs":{},"inventory_items":{},"sessions":{},"schema_version":{}}}"#,
        escape_json_string(campaign_id),
        escape_json_string(&name),
        characters,
        quests_count,
        npcs_count,
        inventory_items,
        sessions_count,
        SCHEMA_VERSION
    );

    respond(stream, 200, &out)
}
